src/uct/cuda/cuda_ipc/cuda_ipc_iface.c

```diff
@@ -96,53 +98,215 @@ static ucs_status_t uct_cuda_ipc_iface_query(uct_iface_h iface,
+    unsigned count = 0;
+    uct_cuda_ipc_event_desc_t *cuda_ipc_event;
+    ucs_queue_iter_t iter;
+    ucs_status_t status;
+
+    ucs_queue_for_each_safe(cuda_ipc_event, iter, event_q, queue) {
+        status = UCT_CUDADRV_FUNC(cuEventQuery(cuda_ipc_event->event));
+        if (UCS_INPROGRESS == status) {
+            continue;
+        }
+        else if (UCS_OK != status) {
+            return status;
+        }
+        ucs_queue_del_iter(event_q, iter);
+        if (cuda_ipc_event->comp != NULL) {
+            uct_invoke_completion(cuda_ipc_event->comp, UCS_OK);
+        }
+        ucs_trace_poll("CUDA_IPC Event Done :%p", cuda_ipc_event);
+        ucs_mpool_put(cuda_ipc_event);
+        count++;
+        if (count >= max_events) {
+            break;
+        }
+    }
+    return count;
+}
+
+static unsigned uct_cuda_ipc_iface_progress(uct_iface_h tl_iface)
+{
+    uct_cuda_ipc_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_ipc_iface_t);
+    unsigned max_events         = iface->config.max_poll;
+
+    return uct_cuda_ipc_progress_event_q(&iface->outstanding_d2d_event_q,
+                                         max_events);
+}
+
 static uct_iface_ops_t uct_cuda_ipc_iface_ops = {
     .ep_get_zcopy             = uct_cuda_ipc_ep_get_zcopy,
     .ep_put_zcopy             = uct_cuda_ipc_ep_put_zcopy,
+    .ep_pending_add           = ucs_empty_function_return_busy,
+    .ep_pending_purge         = ucs_empty_function,
     .ep_flush                 = uct_base_ep_flush,
     .ep_fence                 = uct_base_ep_fence,
     .ep_create_connected      = UCS_CLASS_NEW_FUNC_NAME(uct_cuda_ipc_ep_t),
     .ep_destroy               = UCS_CLASS_DELETE_FUNC_NAME(uct_cuda_ipc_ep_t),
-    .iface_flush              = uct_base_iface_flush,
+    .iface_flush              = uct_cuda_ipc_iface_flush,
     .iface_fence              = uct_base_iface_fence,
-    .iface_progress_enable    = ucs_empty_function,
-    .iface_progress_disable   = ucs_empty_function,
-    .iface_progress           = ucs_empty_function_return_zero,
+    .iface_progress_enable    = uct_base_iface_progress_enable,
+    .iface_progress_disable   = uct_base_iface_progress_disable,
+    .iface_progress           = uct_cuda_ipc_iface_progress,
     .iface_close              = UCS_CLASS_DELETE_FUNC_NAME(uct_cuda_ipc_iface_t),
     .iface_query              = uct_cuda_ipc_iface_query,
     .iface_get_device_address = uct_cuda_ipc_iface_get_device_address,
     .iface_get_address        = uct_cuda_ipc_iface_get_address,
     .iface_is_reachable       = uct_cuda_ipc_iface_is_reachable,
 };
 
+static void uct_cuda_ipc_event_desc_init(ucs_mpool_t *mp, void *obj, void *chunk)
+{
+    uct_cuda_ipc_event_desc_t *base = (uct_cuda_ipc_event_desc_t *) obj;
+    ucs_status_t status;
+
+    memset(base, 0 , sizeof(*base));
+    status = UCT_CUDADRV_FUNC(cuEventCreate(&(base->event),
+                                            CU_EVENT_DISABLE_TIMING));
+    if (UCS_OK != status) {
+        return;
```

Comment:
why is this branch needed? maybe print some error instead of return? The same is relevant for `uct_cuda_ipc_event_desc_cleanup`

Reply:
Actually, UCT_CUDADRV_FUNC itself would print an error. I prefer to return fatal error if events can't be created 
@bureddy